src/encrypt.rs

```diff
@@ -0,0 +1,445 @@
+
+        let iv = match self.unprotected.get(&CborValue::Integer(5)) {
+            Some(CborValue::Bytes(val)) => val,
+            _ => {
+                return Err(COSEError::SpecificationError(
+                    "Unprotected Header contains invalid IV specification".to_string(),
+                ))
+            }
+        };
+
+        let (ciphertext, tag) = self
+            .ciphertext
+            .split_at(self.ciphertext.len() - cose_alg.tag_size());
+
+        let payload = decrypt_aead(
+            cose_alg.openssl_cipher(),
+            key,
+            Some(&iv),
+            &enc_structure
+                .as_bytes()
+                .map_err(COSEError::SerializationError)?,
+            ciphertext,
+            &tag,
+        )
+        .map_err(COSEError::EncryptionError)?;
+
+        Ok((protected, &self.unprotected, payload))
+    }
+
+    /// Serializes the structure for transport / storage. If `tagged` is true, the optional #6.16
+    /// tag is added to the output.
+    pub fn as_bytes(&self, tagged: bool) -> Result<Vec<u8>, COSEError> {
+        let bytes = if tagged {
+            serde_cbor::to_vec(&serde_cbor::tags::Tagged::new(Some(16), &self))
+        } else {
+            serde_cbor::to_vec(&self)
+        };
+        bytes.map_err(COSEError::SerializationError)
+    }
+
+    /// This function deserializes the structure, but doesn't check the contents for correctness
+    /// at all. Accepts untagged structures or structures with tag 16.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, COSEError> {
+        let coseencrypt0: serde_cbor::tags::Tagged<Self> =
+            serde_cbor::from_slice(bytes).map_err(COSEError::SerializationError)?;
+
+        match coseencrypt0.tag {
+            None | Some(16) => (),
+            Some(tag) => return Err(COSEError::TagError(Some(tag))),
+        }
+        let protected = coseencrypt0.value.protected.as_slice();
+        let _: HeaderMap =
+            serde_cbor::from_slice(protected).map_err(COSEError::SerializationError)?;
+        Ok(coseencrypt0.value)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_encrypt_decrypt() {
+        let key = b"\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0A\x0B\x0C\x0D\x0E\x0F";
+        let plaintext = b"\x12\x34\x56\x78\x90\x12\x34\x56\x12\x34\x56\x78\x90\x12\x34\x56";
+        let cencrypt0 = COSEEncrypt0::new(plaintext, CipherConfiguration::Gcm, key).unwrap();
+        let (_, _, dec) = cencrypt0.decrypt(key).unwrap();
+        assert_eq!(dec, plaintext);
+        assert_ne!(
+            plaintext.to_vec(),
+            serde_cbor::to_vec(&cencrypt0.ciphertext).unwrap()
+        );
+        let fromb = COSEEncrypt0::from_bytes(&cencrypt0.as_bytes(true).unwrap()[..]).unwrap();
+        let (_, _, dec) = fromb.decrypt(key).unwrap();
+        assert_eq!(dec, plaintext);
+        assert_ne!(
+            plaintext.to_vec(),
+            serde_cbor::to_vec(&cencrypt0.ciphertext).unwrap()
```

Comment:
Should be "fromb" instead of "cencrypt0" this time.

Reply:
fixed thanks!